src/coreComponents/physicsSolvers/fluidFlow/kernels/compositional/HU2PhaseFlux.hpp

```diff
@@ -0,0 +1,813 @@
+                                                         sesri,
+                                                         sei,
+                                                         potDiff,
+                                                         phaseMob,
+                                                         dPhaseMob,
+                                                         k_up_i,
+                                                         mob_i,
+                                                         dMob_i_dP,
+                                                         dMob_i_dC );
+        localIndex k_up_j = -1;
+        real64 mob_j{};
+        real64 dMob_j_dP{};
+        real64 dMob_j_dC[numComp]{};
+        upwindMobility< numComp, numFluxSupportPoints >( jp,
+                                                         seri,
+                                                         sesri,
+                                                         sei,
+                                                         -potDiff,
+                                                         phaseMob,
+                                                         dPhaseMob,
+                                                         k_up_j,
+                                                         mob_j,
+                                                         dMob_j_dP,
+                                                         dMob_j_dC );
+
+        real64 const mobTot = LvArray::math::max( mob_i + mob_j, minTotMob );
+        real64 const mobTotInv = 1 / mobTot;
+        real64 dMobTot_dP[numFluxSupportPoints]{};
+        real64 dMobTot_dC[numFluxSupportPoints][numComp]{};
+        dMobTot_dP[k_up_i] += dMob_i_dP;
+        dMobTot_dP[k_up_j] += dMob_j_dP;
+        for( localIndex jc = 0; jc < numComp; ++jc )
+        {
+          dMobTot_dC[k_up_i][jc] += dMob_i_dC[jc];
+          dMobTot_dC[k_up_j][jc] += dMob_j_dC[jc];
+        }
+
+        // Assembling gravitational flux phase-wise as \phi_{i,g} = \sum_{k\nei} \lambda_k^{up,g} f_k^{up,g} (G_i - G_k)
+        gravPhaseFlux += mob_i * mob_j * mobTotInv * potDiff;
+
+        // mob_i derivatives
+        dGravPhaseFlux_dP[k_up_i] += dMob_i_dP * mob_j * mobTotInv * potDiff;
+        for( localIndex jc = 0; jc < numComp; ++jc )
+        {
+          dGravPhaseFlux_dC[k_up_i][jc] += dMob_i_dC[jc] * mob_j * mobTotInv * potDiff;
+        }
+
+        // mob_j derivatives
+        dGravPhaseFlux_dP[k_up_j] += mob_i * dMob_j_dP * mobTotInv * potDiff;
+        for( localIndex jc = 0; jc < numComp; ++jc )
+        {
+          dGravPhaseFlux_dC[k_up_j][jc] += mob_i * dMob_j_dC[jc] * mobTotInv * potDiff;
+        }
+
+        // mobTot derivatives
+        real64 const mobTotInv2 = mobTotInv * mobTotInv;
+        for( localIndex ke = 0; ke < numFluxSupportPoints; ++ke )
+        {
+          dGravPhaseFlux_dP[ke] -= mob_i * mob_j * dMobTot_dP[ke] * mobTotInv2 * potDiff;
```

Comment:
i am confused, sorry
it is jsut taking derivatives piece by piece, did i miss something?

Reply:
No worries, just noticing that as you are force to go _fractional flow_ formulation, it might be simpler to pre-compute fractional flow derivatives correctly upwinded so ``d(mob_beta * f_alpha) = d(mob_beta) f_alpha + d(f_alpha) mob_beta`` rather than having fractional flow all explicited. You'll have total mobility correctly upwinded hidden there though. (which is great when everything is working, less then there is a bug to track)
Another good point is that it would retain which phase is considered and which is anonymous iterative index.

Reply:
i got it now, yeah, this is some matter of personal preference i guess
not sure i prefer to introduce another frac flow containers